### ISSUE TITLE
javascript.statements.async_function - Firefox v53 needed not v52

### ISSUE THREAD

**Issue by TheMetaGuy** [mdn/browser-compat-data#20269]
### What type of issue is this?

Incorrect support data (example: Chrome says "86" but support was added in "40")

### What information was incorrect, unhelpful, or incomplete?

The compatibility table states that Firefox v52 supports async functions but V53 appears to be needed.

### What browsers does this problem apply to, if applicable?

Firefox

### What did you expect to see?

v53 not v52 for the minimum version of Firefox required for full async function support

### Did you test this? If so, how?

I am using this function to determine if an object is a async function.  
function isAsyncFunction(obj) {
    if (obj.constructor === undefined) {
        return false;
    }
    if (obj.constructor.name !== "AsyncFunction") {
        return false;
    }
    if (obj.prototype && obj.prototype.then === undefined) {
        return false;
    }
    return true;
}

It only returns true for  FireFox v53 or newer

### MDN URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

### MDN metadata

<details>
<summary>MDN page report details</summary>

* Query: `javascript.statements.async_function`
* Report started: 2023-06-30T16:29:07.229Z

</details>

**Comment by Josh-Cena** [comment]
If `async function` doesn't throw a syntax error, it means the feature is at least partially supported. It would be helpful if you could provide which condition actually fails on FF52.

**Comment by TheMetaGuy** [comment]
Hi Joshua

 

Ok so this is the test which shows the issue on Firefox V52

 

it('get extendedTypeOf  asyncfunction', () => {

    let myAsyncFunction = async function() {};

    if (  isAsyncFunction(myAsyncFunction)  ) {

        expect(extendedTypeOf( myAsyncFunction )).toBe('AsyncFunction');

    } else {

        expect(extendedTypeOf( myAsyncFunction )).toBe('function');  // if called from ES5 code

        console.log('Async functions are not supported in this environment');

    }

});

 

Where 

 

function isAsyncFunction(obj) {

    if (obj.constructor === undefined) {

        return false;

    }

    if (obj.constructor.name !== "AsyncFunction") {

        return false;

    }

    if (obj.prototype && obj.prototype.then === undefined) {

        return false;

    }

    return true;

}

 

My extendedTypeOf( )  is part of a small library I’ve put on NPM - https://www.npmjs.com/package/type-master . It’s basically an improved version of the built-in  ‘typeof’ in that it actually returns accurate type information for non-primitives, ES6 types etc  

 

Whilst testing across old browsers I found I needed to add the isAsyncFunction( ) as a test because the ‘type’ returned for async functions on FF v52 is ‘function’.  

 

My extendedTypeOf( ) uses the Object.prototype.tostring( ) behind the scenes ( along with some other checks) and suspect the Issue is that FF v52 may provide the correct functionality for ‘async functions’ (I.e behaviour is OK) but the underlying ‘type’ is still that of a ‘function’.   

 

In a nutshell I think 

 

// Firefox v52

console.log( Object.prototype.toString.call( async function() { }) ); // [object Function]

 

// Firefox v53+

console.log( Object.prototype.toString.call( async function() { }) ); // [object AsyncFunction]

 

Regards 

 

Owen

**Comment by Josh-Cena** [comment]
> console.log( Object.prototype.toString.call( async function() { }) ); // [object Function]

Did you test this? I tested on Firefox 52, and did not get this behavior.

<img width="1194" alt="image" src="https://github.com/mdn/browser-compat-data/assets/55398995/05e9f62b-e3ce-4199-8ccb-d66641250012">

**Comment by TheMetaGuy** [comment]
Hi Joshua

 

Apologies.  You are correct.   I have re-installed FF v52 and   Object.prototype.toString.call( async function() { } ));   gives  ==>  [object AsyncFunction]

 

I have naively been relying on this Babel setting in running part of my unit tests

 

    "babel": {

        "presets": [

            [

                ***@***.***/preset-env",

                {

                    "targets": {

                        "browsers": [

                            "Edge >= 15",

                            "Chrome >= 55",

                            "Firefox >= 53",

                            "Opera >= 43",

                            "Safari >= 10.2",

                            "iOS >= 11.2",

                            "ChromeAndroid >= 55",

                            "Samsung >= 6.0"

                        ]

                    }

                }

            ]

        ]

    },

 

And I get the problem if the FireFox line is ‘>=52’  but not if it’s  ‘>=53’  So they problem is with Babel and the version of Firefox it is actually set for as opposed to what is specfied 

 

Apologies again for my error.  I should have manually checked inside the browser instead of relying on what Babel says it emulating / using.

 

Regards

 

Owen

**Comment by TheMetaGuy** [comment]
This can be closed.  It actually was a Babel preset issue When set for Firefox 52 it actually using / emulating a lower version

**Comment by TheMetaGuy** [comment]
I have closed this issue. Apologies again 

 

Owen

 

From: Owen Cullum ***@***.***> 
Sent: 04 July 2023 00:31
To: 'mdn/browser-compat-data' ***@***.***>; 'mdn/browser-compat-data' ***@***.***>
Cc: 'Author' ***@***.***>
Subject: RE: [mdn/browser-compat-data] javascript.statements.async_function - Firefox v53 needed not v52 (Issue #20269)

 

Hi Joshua

 

Apologies.  You are correct.   I have re-installed FF v52 and   Object.prototype.toString.call( async function() { } ));   gives  ==>  [object AsyncFunction]

 

I have naively been relying on this Babel setting in running part of my unit tests

 

    "babel": {

        "presets": [

            [

                ***@***.***/preset-env",

                {

                    "targets": {

                        "browsers": [

                            "Edge >= 15",

                            "Chrome >= 55",

                            "Firefox >= 53",

                            "Opera >= 43",

                            "Safari >= 10.2",

                            "iOS >= 11.2",

                            "ChromeAndroid >= 55",

                            "Samsung >= 6.0"

                        ]

                    }

                }

            ]

        ]

    },

 

And I get the problem if the FireFox line is ‘>=52’  but not if it’s  ‘>=53’  So they problem is with Babel and the version of Firefox it is actually set for as opposed to what is specfied 

 

Apologies again for my error.  I should have manually checked inside the browser instead of relying on what Babel says it emulating / using.

 

Regards

 

Owen
